### PR TITLE
Merge branch 'rene_WSIWindow_and_Teapots'

### PR DESCRIPTION
**Pull request by renelindsay**
WSIWindow Utility for creating a Vulkan Window.
Provides event callbacks for mouse, keyboard and multi-touch events.
Runs on Windows, Linux and Android.

Change-Id: I8019075d3ba7e44cfaf78f6b242cc20000000016

### CODE REVIEW

**Comment by olvaffe** [comment]
I am fine with the change, as long as it does not break anything.

File: Utilities/WSIWindow/Validation.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#endif
+
+//-------------------------------------------------------------
+
+//-----------------------Error Checking------------------------
+#if !defined(NDEBUG) || defined(ENABLE_LOGGING) || defined(ENABLE_VALIDATION)
+//  In Debug mode, convert a VkResult return value to a string.
+const char* VkResultStr(VkResult err){
+    switch (err) {
+#define STR(r) case r: return #r
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/Validation.cpp]: Inline this code -- no need for macros here.

File: Utilities/WSIWindow/Validation.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        }
+//--------------------------------------------------------------------------------------------
+
+//----------------------------------------CDebugReport----------------------------------------
+
+void CDebugReport::Init(VkInstance inst){
+    assert(!!inst);
+    GET_INSTANCE_PROC_ADDR(inst, CreateDebugReportCallbackEXT);
+    GET_INSTANCE_PROC_ADDR(inst,DestroyDebugReportCallbackEXT);
+  //GET_INSTANCE_PROC_ADDR(inst,       DebugReportMessageEXT );
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/Validation.cpp]: kill

File: Utilities/WSIWindow/Validation.h [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+#if defined(__ANDROID__) && !defined(NDEBUG)
+  #define ENABLE_VALIDATION 1
+  #define ENABLE_LOGGING    1
+  //#define NDEBUG               //cuts 4kb off apk size
+#endif
+//------------------------------------------
+//===========================================Check VkResult=============================================
+// Macro to check VkResult for errors(negative) or warnings(positive), and print as a string.
+#ifdef NDEBUG                           //In release builds, don't print VkResult strings.
+  #define VKERRCHECK(VKFN) { VKFN; }
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/Validation.h]: Convert this to a function.

File: Utilities/WSIWindow/Validation.h [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    #define cTICK "\u2713"
+#elif __LINUX__
+    #include <xkbcommon/xkbcommon.h>
+    #define ANSICODE(x) x     /* Enable ANSI codes on Linux */
+    #define cTICK "\u2713"
+#endif
+//----------------------------------------------------------------------------------
+
+//--- ANSI escape codes to set text colours. eg. printf(cRED"Red text." cRESET); ---
+#define cFAINT     ANSICODE("\033[38;2;96;96;96m")
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/Validation.h]: Can these be static variables?

File: Utilities/WSIWindow/Validation.h [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+#else
+  #define cSTRIKEOUT // linux only
+#endif
+//----------------------------------------------------------------------------------
+//----------------Printing Log & Validation messages on Android vs PC---------------
+#ifdef ANDROID
+    #include <jni.h>
+    #include <android/log.h>
+    #define LOG_TAG    "WSIWindow"
+    #define _LOG(...)    __android_log_print(ANDROID_LOG_INFO   ,LOG_TAG,__VA_ARGS__)
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/Validation.h]: Convert these to functions -- avoid macros.

File: Utilities/WSIWindow/Validation.h [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+const char* VkResultStr(VkResult err);  //Convert vulkan result code to a string.
+void ShowVkResult(VkResult err);        //Print warnings and errors.
+
+//============================================ CDebugReport ============================================
+class CDebugReport{
+    CDebugReport();
+    PFN_vkCreateDebugReportCallbackEXT  vkCreateDebugReportCallbackEXT;
+    PFN_vkDestroyDebugReportCallbackEXT vkDestroyDebugReportCallbackEXT;
+  //PFN_vkDebugReportMessageEXT         vkDebugReportMessageEXT;
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/Validation.h]: kill

File: Utilities/WSIWindow/VulkanWrapper/vulkan/vk_platform.h [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    #define VKAPI_ATTR
+    #define VKAPI_CALL
+    #define VKAPI_PTR
+#endif
+
+#include <stddef.h>
+
+#if !defined(VK_NO_STDINT_H)
+    #if defined(_MSC_VER) && (_MSC_VER < 1600)
+        typedef signed   __int8  int8_t;
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/VulkanWrapper/vulkan/vk_platform.h]: Is this necessary?  Aren't these already defined for in vk_platform.h?

File: Utilities/WSIWindow/android/native.cpp [hunk 8]
```diff
@@ -0,0 +1,127 @@
+#include "native.h"
+
+android_app* Android_App=0;                //Android native-actvity state
+/*
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/android/native.cpp]: kill this commented out stuff.

File: Utilities/WSIWindow/android/native.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+//------------------------------------------------
+*/
+//====================Main====================
+int main(int argc, char *argv[]);          //Forward declaration of main function
+
+void android_main(struct android_app* state) {
+    printf("Native Activity\n");
+    app_dummy();                           // Make sure glue isn't stripped
+    //state->onAppCmd     = handle_cmd;      // Register window event callback  (Temporary)
```
Comment (mark-lunarg) [inline Utilities/WSIWindow/android/native.cpp]: In general, we should remove any dead code that gets checked into the repo.

File: Utilities/WSIWindow/VulkanWrapper/vulkan/vk_platform.h (hunk 7 above)
Comment (renelindsay) [inline Utilities/WSIWindow/VulkanWrapper/vulkan/vk_platform.h]: This IS vk_platform.h.  
I haven't made any changes to this file.

File: Utilities/WSIWindow/Validation.h (hunk 6 above)
Comment (renelindsay) [inline Utilities/WSIWindow/Validation.h]: done

File: Utilities/WSIWindow/android/native.cpp (hunk 8 above)
Comment (renelindsay) [inline Utilities/WSIWindow/android/native.cpp]: The commented out function is a work in progress, and not ready yet.
A few Android event types are not being handled yet.

**Pull request closed without merging**